packages/core/src/dispatcher.ts

```diff
@@ -138,40 +159,28 @@ export class Dispatcher {
    * @param streamId - StreamID of the stream the commit belongs to, used for logging.
    */
   async storeCommit(data: any, streamId?: StreamID): Promise<CID> {
+    const carFile = carFactory.build()
     Metrics.count(COMMITS_STORED, 1)
     try {
       if (StreamUtils.isSignedCommitContainer(data)) {
         const { jws, linkedBlock, cacaoBlock } = data
         // if cacao is present, put it into ipfs dag
         if (cacaoBlock) {
           const decodedProtectedHeader = base64urlToJSON(data.jws.signatures[0].protected)
-          const capIPFSUri = decodedProtectedHeader.cap
-          await this._shutdownSignal.abortable((signal) => {
-            return Utils.putIPFSBlock(capIPFSUri, cacaoBlock, this._ipfs, signal)
-          })
+          const capCID = CID.parse(decodedProtectedHeader.cap.replace('ipfs://', ''))
+          carFile.blocks.put(new CarBlock(capCID, cacaoBlock))
         }
 
-        // put the JWS into the ipfs dag
-        const cid = await this._shutdownSignal.abortable((signal) => {
-          return this._ipfs.dag.put(jws, {
-            storeCodec: 'dag-jose',
-            hashAlg: 'sha2-256',
-            signal: signal,
-          })
-        })
-        // put the payload into the ipfs dag
-        const linkCid = jws.link
-        await this._shutdownSignal.abortable((signal) => {
-          return Utils.putIPFSBlock(linkCid, linkedBlock, this._ipfs, signal)
-        })
-        await this._restrictCommitSize(jws.link.toString())
-        await this._restrictCommitSize(cid)
+        carFile.blocks.put(new CarBlock(jws.link, linkedBlock)) // Encode payload
+        restrictCommitBlockSize(carFile, jws.link)
+        const cid = carFile.put(jws, { codec: 'dag-jose', hasher: 'sha2-256', isRoot: true }) // Encode JWS itself
+        restrictCommitBlockSize(carFile, cid)
+        await all(this._ipfs.dag.import(carFile, { pinRoots: false }))
         return cid
       }
-      const cid = await this._shutdownSignal.abortable((signal) => {
-        return this._ipfs.dag.put(data, { signal: signal })
-      })
-      await this._restrictCommitSize(cid)
+      const cid = carFile.put(data)
```

Comment:
Shouldn't this be the root?
```suggestion
      const cid = carFile.put(data, { isRoot: true })
```